[RFC] Decouple fast pass functions

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46469 [RFC] Decouple fast pass functions**

There are some "fast_pass" function calls, where the symbols in `ATen/native` are directly referenced from outside of native at linking stage. This PR is to decouple one of the fast pass from native, while keeping the same functionality. `scalar_to_tensor` is included through `ATen/ATen.h`, which could be referenced by any cpp file including this header. 

This PR is prerequisite of per-app selective build on `ATen/native`.

Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

iseeyuan committed:
[RFC] Decouple fast pass functions



Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[RFC] Decouple fast pass functions"


Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]

smessmer reviewed (approved):
Looks good to me, but I'd like @ezyang to take a look since he's spend more time thinking about fast pass recently than I did and I think there's an issue somewhere about allowing fast pass for ops in native_functions.yaml.

smessmer commented:
got one question while re-reading this. Do we want to add `_fast` functions for all our ops? Or should we just define the direct op without suffix (i.e. at::scalar_tensor in this case) in the FastPass.cpp file?

ezyang commented:
I don't understand why this works. My understanding is you need to prevent direct symbol references from ATen (base dir) into ATen_cpu (native/ dir) so that you can compile ATen without direct reference to ATen_cpu. Fine. But FastPath.cpp still refers to a native:: function. 

I'd also examine whether or not the "FIXME" is still true or not; if it's not true anymore, you could probably just replace this with a call to a dispatched function and call it a day.

iseeyuan commented:
> got one question while re-reading this. Do we want to add `_fast` functions for all our ops? Or should we just define the direct op without suffix (i.e. at::scalar_tensor in this case) in the FastPass.cpp file?

No specific reason on "_fast". I was thinking of "at::scalar_tensor", but it's already used.

iseeyuan commented:
> I don't understand why this works. My understanding is you need to prevent direct symbol references from ATen (base dir) into ATen_cpu (native/ dir) so that you can compile ATen without direct reference to ATen_cpu. Fine. But FastPath.cpp still refers to a native:: function.
Good catch! I should also move at::native::empty_cpu out of native, which can be done relatively easily, because it does not reference to other native symbols.

> 
> I'd also examine whether or not the "FIXME" is still true or not; if it's not true anymore, you could probably just replace this with a call to a dispatched function and call it a day.

"FIXME" suggests to implement it in Scalar::toTensor, but I couldn't find "toTensor" in Scalar. I don't understand "but there is currently no way to implement this without going through Derived Types (which are not part of core)". Does it mean the derived types of Scalar or derived type of Tensor?

If I debug the very original code (without @ailzhang 's perf change), it calls "at::scalar_tensor", which did nothing but dispatch to "aten::scalar_tensor", and eventually calls at::native::scalar_tensor. I believe @ailzhang added the fast pass change because it's faster than at::scalar_tensor. If now at::scalar_tensor is as fast, that would be an easy fix! @ailzhang , could you please confim?

dhruvbird commented on aten/src/ATen/ScalarOps.h:
```diff
@@ -11,16 +12,16 @@ namespace c10 {
-  if (device == at::kCPU) {
+  if (device == at::kCPU && !s.isComplex()) {
     if (s.isFloatingPoint()) {
-      return at::native::scalar_tensor(s, at::device(at::kCPU).dtype(at::kDouble));
+      return at::scalar_tensor_fast(s, at::device(at::kCPU).dtype(at::kDouble));
     } else if (s.isBoolean()) {
-      return at::native::scalar_tensor(s, at::device(at::kCPU).dtype(at::kBool));
+      return at::scalar_tensor_fast(s, at::device(at::kCPU).dtype(at::kBool));
     } else if (s.isComplex()) {
```
Q. @iseeyuan Maybe I'm missing something, but the conditions here seem to first exclude any scalar that is complex and then later checks if a complex scalar was found.

iseeyuan replied:
Good catch. The later complex check part should be removed because it's not supported by fast pass.

dhruvbird commented on aten/src/ATen/ScalarOps.h:
```diff
@@ -11,16 +12,16 @@ namespace c10 {
-  if (device == at::kCPU) {
+  if (device == at::kCPU && !s.isComplex()) {
     if (s.isFloatingPoint()) {
-      return at::native::scalar_tensor(s, at::device(at::kCPU).dtype(at::kDouble));
```
Q. @iseeyuan How was the existing code calling into aten::native w/o including the relevant headers first? (just curious - maybe I'm missing something here).

ljk53 replied:
all native functions are declared in this generated header: https://codebrowser.bddppq.com/pytorch/pytorch/build/aten/src/ATen/NativeFunctions.h.html

ljk53 commented on aten/src/ATen/FastPass.cpp:
```diff
@@ -0,0 +1,34 @@
+// FastPass
```
You mentioned there are some "fast_pass" function calls - do you know roughly how many of them?
Are we going to copy all their cpu implementation into this file? How is it compared to putting it in ScalarOps.cpp?

iseeyuan replied:
scalar_tensor is the one referenced from lite trainer. There can be others but likely not built for mobile. It's a good idea to put it in ScalarOps.cpp!

smessmer commented:
> > got one question while re-reading this. Do we want to add `_fast` functions for all our ops? Or should we just define the direct op without suffix (i.e. at::scalar_tensor in this case) in the FastPass.cpp file?
> 
> No specific reason on "_fast". I was thinking of "at::scalar_tensor", but it's already used.

Right but isn't `at::scalar_tensor_fast` doing the same as `at::scalar_tensor`? I thought you're just trying to move `at::scalar_tensor` to a different compilation unit. Maybe I misunderstood your PR.

ezyang commented:
> Good catch! I should also move at::native::empty_cpu out of native, which can be done relatively easily, because it does not reference to other native symbols.

OK, I think I understand the basic thrust of "fast path", however, I'm going to suggest we name them differently. What we are talking about here is a sort of "core" set of operators/operator-like-functionality that is *always* compiled into your ATen binary, as opposed to being dynamically loadable via ATen_cpu. The fact that their presence in ATen allows for fastpathing is a happy side effect, but it's not the primary effect (and in a lot of cases, it's not permissible to fastpath in the same way suggested in https://github.com/pytorch/pytorch/issues/30774 -- for example, at::empty cannot fastpath to at::native::empty_cpu, because it needs to work for other devices rather than CPU). Maybe "primitive" or "builtins" or "wired" in operators are better names for them.

Some other thoughts:

* These ops shouldn't get duplicated with their versions in native/; instead, the native/ implementation should just call the version of the operator in ATen. So for example, in your PR, you have `fill_fast`; what I would expect to see is that the fill implementation in operators calls that. This will help us keep the implementations lined up. And if it turns out that fill in CPU is actually vectorized, so you can't actually replicate the behavior here, then what we're putting here is a simple, "slow" implementation that doesn't need to be compiled multiple times, well, that's something that should be reflected in the naming (and we shouldn't call it fast!!!)
* But, to contradict myself above, I'm OK if you need to unblock selective build on ATen_cpu if you want to just inline a simple implementation of fill and empty *only* for the implementation of `scalar_tensor`. But the idea would be that this duplicate is totally private and not intended to be used for anything else (and not setting up a new pattern of operators in ATen library directly.)

> "FIXME" suggests to implement it in Scalar::toTensor, but I couldn't find "toTensor" in Scalar. I don't understand "but there is currently no way to implement this without going through Derived Types (which are not part of core)". Does it mean the derived types of Scalar or derived type of Tensor?

The FIXME text looks very old. Derived Types refers to things like CPUType/CUDAType. But I half suspect this dates back to the day when these were actual honest to goodness types. Honestly, I'd just ignore it entirely, as you do in your next paragraph.

The fastpath dates to https://github.com/pytorch/pytorch/pull/29915 which has a relevant benchmark.

iseeyuan commented:
> > > got one question while re-reading this. Do we want to add `_fast` functions for all our ops? Or should we just define the direct op without suffix (i.e. at::scalar_tensor in this case) in the FastPass.cpp file?
> > 
> > 
> > No specific reason on "_fast". I was thinking of "at::scalar_tensor", but it's already used.
> 
> Right but isn't `at::scalar_tensor_fast` doing the same as `at::scalar_tensor`? I thought you're just trying to move `at::scalar_tensor` to a different compilation unit. Maybe I misunderstood your PR.
> 
> Since `at::scalar_tensor` is in `native_functions.yaml`, it's generated by our codegen and we'd likely have to disable that if we want to implement it manually in a different compilation unit. That's what I meant above about me remembering that there is an issue about allowing fast pass for ops in native_functions.yaml.

I believe `at::scalar_tensor` is doing slower than `aten::native::scalar_tensor`, otherwise we don't have to have the change in https://github.com/pytorch/pytorch/pull/29915. The change is majorly calling aten::native::scalar_tensor directly for fast pass in CPU, otherwise fall back to at::scalar_tensor. It's hard to keep at::native:scalar_tensor in aten_cpu while moving other native files to app level. What I'm doing is to extract the "fast" part of scalar to tensor, instead of calling the generic Tensor fill. 

If only `at::scalar_tensor` is used, it goes through dispatch to call `aten::native::scalar_tensor`, and there should be no linking error when we move `aten::native::scalar_tensor` to app level.

iseeyuan commented:
> > Good catch! I should also move at::native::empty_cpu out of native, which can be done relatively easily, because it does not reference to other native symbols.
> 
> OK, I think I understand the basic thrust of "fast path", however, I'm going to suggest we name them differently. What we are talking about here is a sort of "core" set of operators/operator-like-functionality that is _always_ compiled into your ATen binary, as opposed to being dynamically loadable via ATen_cpu. The fact that their presence in ATen allows for fastpathing is a happy side effect, but it's not the primary effect (and in a lot of cases, it's not permissible to fastpath in the same way suggested in #30774 -- for example, at::empty cannot fastpath to at::native::empty_cpu, because it needs to work for other devices rather than CPU). Maybe "primitive" or "builtins" or "wired" in operators are better names for them.
> 
> Some other thoughts:
> 
> * These ops shouldn't get duplicated with their versions in native/; instead, the native/ implementation should just call the version of the operator in ATen. So for example, in your PR, you have `fill_fast`; what I would expect to see is that the fill implementation in operators calls that. This will help us keep the implementations lined up. And if it turns out that fill in CPU is actually vectorized, so you can't actually replicate the behavior here, then what we're putting here is a simple, "slow" implementation that doesn't need to be compiled multiple times, well, that's something that should be reflected in the naming (and we shouldn't call it fast!!!)
> * But, to contradict myself above, I'm OK if you need to unblock selective build on ATen_cpu if you want to just inline a simple implementation of fill and empty _only_ for the implementation of `scalar_tensor`. But the idea would be that this duplicate is totally private and not intended to be used for anything else (and not setting up a new pattern of operators in ATen library directly.)
> 
> > "FIXME" suggests to implement it in Scalar::toTensor, but I couldn't find "toTensor" in Scalar. I don't understand "but there is currently no way to implement this without going through Derived Types (which are not part of core)". Does it mean the derived types of Scalar or derived type of Tensor?
> 
> The FIXME text looks very old. Derived Types refers to things like CPUType/CUDAType. But I half suspect this dates back to the day when these were actual honest to goodness types. Honestly, I'd just ignore it entirely, as you do in your next paragraph.
> 
> The fastpath dates to #29915 which has a relevant benchmark.

I'll look into the code and see if some functions can be re-used. I was trying to make the functions be reused as much as possible, but realized changing a scalar to tensor is quite different from the existing flow of a generic empty tensor -> fill. 

The naming suggestion makes sense to me. I'll change the names accordingly.

iseeyuan committed:
Update on "[RFC] Decouple fast pass functions"


There are some "fast_pass" function calls, where the symbols in `ATen/native` are directly referenced from outside of native at linking stage. This PR is to decouple one of the fast pass from native, while keeping the same functionality. `scalar_to_tensor` is included through `ATen/ATen.h`, which could be referenced by any cpp file including this header. 

This PR is prerequisite of per-app selective build on `ATen/native`.

Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[RFC] Decouple fast pass functions"


There are some "fast_pass" function calls, where the symbols in `ATen/native` are directly referenced from outside of native at linking stage. This PR is to decouple one of the fast pass from native, while keeping the same functionality. `scalar_to_tensor` is included through `ATen/ATen.h`, which could be referenced by any cpp file including this header. 

This PR is prerequisite of per-app selective build on `ATen/native`.

Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[RFC] Decouple fast pass functions"


There are some "fast_pass" function calls, where the symbols in `ATen/native` are directly referenced from outside of native at linking stage. This PR is to decouple one of the fast pass from native, while keeping the same functionality. `scalar_to_tensor` is included through `ATen/ATen.h`, which could be referenced by any cpp file including this header. 

This PR is prerequisite of per-app selective build on `ATen/native`.

Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorFactories.h:
```diff
@@ -62,9 +62,7 @@ inline void check_args(
-    TORCH_CHECK(x >= 0, "Trying to create tensor with negative dimension ", x, ": ", size);
-  }
+  at::check_size_nonnegative(size);
 }
```
Can't you just say `using at::check_size_nonnegative` if you don't want to renamespace everything? (However, I kind of feel like C++ namespacing rules should make this declarations 100% unnecessary--maybe there are ADL shenanigans going on)

ezyang commented on aten/src/ATen/Utils.h:
```diff
@@ -120,4 +120,13 @@ static inline T* get_generator_or_default(const c10::optional<Generator>& gen, c
+}
+
+CAFFE2_API
+Tensor empty_cpu(IntArrayRef size, const TensorOptions& options_, c10::optional<c10::MemoryFormat> optional_memory_format);
```
I'm a bit skeptical that these should be dumped in the top level `at::` namespace. They don't seem like user-facing functions, and they have extra preconditions for their use which are not necessarily obvious to users.

iseeyuan replied:
at:: namespace does look like top-level. Should I use existing namespace like at::detail, or just create a new namespace? I think using existing namespace makes sense. Which namespace do you prefer?

ezyang commented on aten/src/ATen/Utils.cpp:
```diff
@@ -12,4 +14,44 @@ int _crash_if_asan(int arg) {
   return x[0];
 }
 
+Tensor empty_cpu(IntArrayRef size, const TensorOptions& options_, c10::optional<c10::MemoryFormat> optional_memory_format) {
```
Probably a good idea to have a little comment here explaining why this lives here and not in `native::`

ezyang commented on aten/src/ATen/ScalarOps.h:
```diff
@@ -4,23 +4,32 @@
+  auto value = value_scalar.to<scalar_t>();
+  scalar_t * dptr = static_cast<scalar_t *>(self.data_ptr());
+  *dptr = value;
+}
```
I don't think this needs to live in the header; it is only used from one cpp file right?

ezyang commented on aten/src/ATen/ScalarOps.cpp:
```diff
@@ -0,0 +1,33 @@
+  TORCH_CHECK(options.device() == kCPU);
+  at::tracer::impl::NoTracerDispatchMode tracer_guard;
+  at::AutoNonVariableTypeMode non_var_type_mode(true);
+  auto result = native::empty_cpu({}, options);
```
You're still going to `native::` namespace here, that violates the library separation invariant right?

iseeyuan replied:
Good catch! I made the implementation without native, but was still calling the native one...

ezyang commented:
The PR is shaping up. My concerns primarily are concerned with:

* There still seem to be layering violations (`at::` calling `at::native::` directly)
* You are accidentally publishing functions to top level namespace (`at::`) when they should be private
* There is not good in code documentation to guide people when they may want to fast path other functions

iseeyuan committed:
Update on "[RFC] Decouple fast pass functions"


There are some "fast_pass" function calls, where the symbols in `ATen/native` are directly referenced from outside of native at linking stage. This PR is to decouple one of the fast pass from native, while keeping the same functionality. `scalar_to_tensor` is included through `ATen/ATen.h`, which could be referenced by any cpp file including this header. 

This PR is prerequisite of per-app selective build on `ATen/native`.

Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]

iseeyuan commented:
> The PR is shaping up. My concerns primarily are concerned with:
> 
> * There still seem to be layering violations (`at::` calling `at::native::` directly)
> * You are accidentally publishing functions to top level namespace (`at::`) when they should be private
> * There is not good in code documentation to guide people when they may want to fast path other functions

Thanks for the comments! Just updated the codes. For namespace I'm using `at::detail`. Please let me know if it's fine.

ezyang commented on aten/src/ATen/ScalarOps.h:
```diff
@@ -4,23 +4,26 @@
 
+namespace at {
+Tensor& scalar_fill(Tensor& self, Scalar value);
+TORCH_API Tensor scalar_tensor_static(Scalar s, const TensorOptions& options);
```
Should these be in top level namespace too?

ezyang replied:
If they are, they better have comments saying what they do! (But I suspect this was not intentional)

iseeyuan replied:
Moved them to at::detail.

ezyang commented on aten/src/ATen/ScalarOps.cpp:
```diff
@@ -0,0 +1,41 @@
+Tensor scalar_tensor_static(Scalar s, const TensorOptions& options) {
+  TORCH_CHECK(options.device() == kCPU);
+  at::tracer::impl::NoTracerDispatchMode tracer_guard;
+  at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Wondering if these guards are necessary. You're calling straight to `detail::empty_cpu` so it doesn't seem like you should hit dispatcher anyway.

ezyang reviewed (approved):
Comments are still missing, but this is good enough to get a preapprove

iseeyuan committed:
Update on "[RFC] Decouple fast pass functions"


There are some "fast_pass" function calls, where the symbols in `ATen/native` are directly referenced from outside of native at linking stage. This PR is to decouple one of the fast pass from native, while keeping the same functionality. `scalar_to_tensor` is included through `ATen/ATen.h`, which could be referenced by any cpp file including this header. 

This PR is prerequisite of per-app selective build on `ATen/native`.

Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[RFC] Decouple fast pass functions"


There are some "fast_pass" function calls, where the symbols in `ATen/native` are directly referenced from outside of native at linking stage. This PR is to decouple one of the fast pass from native, while keeping the same functionality. `scalar_to_tensor` is included through `ATen/ATen.h`, which could be referenced by any cpp file including this header. 

This PR is prerequisite of per-app selective build on `ATen/native`.

Differential Revision: [D24361863](https://our.internmc.facebook.com/intern/diff/D24361863/)

[ghstack-poisoned]
